packages/ui/src/components/form-primitives/textarea.tsx

```diff
@@ -0,0 +1,130 @@
+import { forwardRef, TextareaHTMLAttributes, useCallback, useMemo, useState } from 'react'
+
+import { ControlGroup, FormCaption, Label } from '@/components'
+import { anyTypeOf, generateAlphaNumericHash, useMergeRefs } from '@/utils'
+import { cn } from '@utils/cn'
+import { cva, VariantProps } from 'class-variance-authority'
+
+const textareaVariants = cva('cn-textarea', {
+  variants: {
+    theme: {
+      default: '',
+      danger: 'cn-textarea-danger',
+      warning: 'cn-textarea-warning'
+    }
+  },
+  defaultVariants: {
+    theme: 'default'
+  }
+})
+
+export interface TextareaProps extends TextareaHTMLAttributes<HTMLTextAreaElement> {
+  label?: string
+  theme?: VariantProps<typeof textareaVariants>['theme']
+  caption?: string
+  error?: string
+  warning?: string
+  optional?: boolean
+  resizable?: boolean
+  maxCharacters?: number
+}
+
+/**
+ * A forward-ref Textarea component with support for labels, captions, and error messages.
+ * @example
+ * <Textarea name="textarea" label="Textarea" placeholder="Enter text here" />
+ */
+const Textarea = forwardRef<HTMLTextAreaElement, TextareaProps>(
+  (
+    {
+      id: defaultId,
+      label,
+      error,
+      warning,
+      caption,
+      optional,
+      disabled,
+      onChange,
+      className,
+      maxCharacters,
+      resizable = false,
+      ...props
+    },
+    ref
+  ) => {
+    const [counter, setCounter] = useState(0)
+
+    const theme = error ? 'danger' : warning ? 'warning' : props.theme
+
+    const id = useMemo(() => defaultId || `textarea-${generateAlphaNumericHash(10)}`, [defaultId])
+
+    const setCharactersCount = useCallback(
+      (value: string) => {
+        if (maxCharacters === undefined) return
+
+        setCounter(value.length)
+      },
+      [maxCharacters]
+    )
+
+    const handleChange = (e: React.ChangeEvent<HTMLTextAreaElement>) => {
+      setCharactersCount(e.target.value)
+      onChange?.(e)
+    }
+
+    const mergedRef = useMergeRefs<HTMLTextAreaElement>([
+      node => {
+        if (!node) return
+
+        if (anyTypeOf(node.value, ['number', 'string'])) {
+          setCharactersCount(String(node.value))
+        }
+      },
+      ref
+    ])
+
+    return (
+      <ControlGroup>
+        <div className="cn-textarea-label-wrapper">
+          {label && (
+            <Label disabled={disabled} optional={optional} htmlFor={id}>
+              {label}
+            </Label>
+          )}
+
+          {maxCharacters && (
+            <span
+              className={cn('cn-textarea-counter', {
+                'cn-textarea-counter-danger': theme === 'danger',
```

Comment:
We might not want to do this based on the `theme`. Some usages won't be in a scenario where there is validation triggering `error` messages. And there will be cases where the validation has nothing to do with the count. I'd change this to be based entirely on the count exceeding the max instead.
```suggestion
                'cn-textarea-counter-danger': counter > maxCharacters,
```

Reply:
Your comment made me think of what if we want a user to write a comment/description of no less than 40 characters - in this case, the counter should be in `"error"` state until it's `40/40`.

I understand your point that some errors cannot be related to the counter, but making the counter too smart or too complex in terms of validation can lead to a worse DX, imo.

Reply:
Oooohhhh, good point. So we might need a `minCharaters` prop that addresses that use-case. So with `maxCharacters` we can show `danger` with `counter > maxCharacters`, and with `minCharacters` we can show `danger` with `counter < minCharacters`. We might also want to think of how we can handle when it's not within the `minCharacters` and `maxCharacters`, but I think we'll need a different visual treatment for that.

Reply:
We can perhaps revisit as an enhancement and just keep it super-dumb without `cn-textarea-counter-danger` for now

Reply:
I would rather do the second small iteration when we decide on the behavior and API

Reply:
Removed `cn-textarea-counter-danger`